### ISSUE TITLE
Out of memory error when parsing large files

### ISSUE THREAD

**Issue by Genubath** [nickbabcock/jomini#38]
Currently, I am getting `FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory` when the output of a large savegame file is passed into `jomini.parse()`
The save file I'm using is about 4.3 million lines 😝 

Implementation is up to you, but maybe there is a way to chunk the parsing operation?

**Comment by nickbabcock** [comment]
Yeah I've noticed this too 😞 

Perchance is this an EU4 save file? I'm in the process of revamping the parser here to handle 7 million line eu4 save files.  So if you're working with EU4, I can link a preview so you can tell me if it works or not. 

> maybe there is a way to chunk the parsing operation?

I think the answer is technically yes, but practically no. I've found that the whole document has to live in memory as it can be ambiguous whether a member is scalar (single value), array, or an object and so sometimes one has to refer to other parts of the document to figure this out. For instance, you can have

```
ENG={
  army={
    name="1st army"
  }
  navy={
    name="1st navy"
  }
}
FRA={
  army={
    name="1st army"
  }
  navy={
    name="1st navy"
  }
  army={
    name="2nd army"
  }
}
```

In this case parsing ENG first would give the false impression that `army` is just an object until you reach FRA where you see that it is an array and that it can appear out of order. Chunking works best when the grammar of everything is known ahead of time

**Comment by Genubath** [comment]
I actually happen to be using a CK2 save file. Perhaps you can do a pass through the file and count the number of elements in between two brackets to determine if they should be an object or an array and store that in an array to refer to when you actually do the parse operation.

**Comment by nickbabcock** [comment]
> I actually happen to be using a CK2 save file

Not problem, just curious as to the files people are parsing 👌 

> Perhaps you can do a pass through the file and count the number of elements in between two brackets to determine if they should be an object or an array and store that in an array to refer to when you actually do the parse operation.

I think that is a decent idea. I'd definitely have to do some tests to see if there are any tradeoffs.

Just so I have a handle on your use case. For CK2 saves, are you searching for a small subset of the save or just looking at the data in general?

**Comment by Genubath** [comment]
I'm just looking at the data in general.

**Comment by nickbabcock** [comment]
Please try installing the `@next` package.

```
npm i jomini@next
```

It contains the new engine. Below is some sample code to get you started:

```js
const { readFileSync, writeFileSync } = require("fs");
const { Jomini } = require(".");
const args = process.argv.slice(2);
if (args.length !== 1) {
  console.info("expected one argument for the file path");
  process.exit(1);
}

console.time("read file");
const buffer = readFileSync(args[0]);
console.timeLog("read file");

(async function () {
  let start = process.hrtime();
  console.time("initialize jomini");
  const parser = await Jomini.Initialize();
  console.timeLog("initialize jomini");

  console.time("parse text");
  const data = parser.parseText(buffer);
  console.timeLog("parse text");

  console.time("stringify json");
  const json = JSON.stringify(data, null, 2);
  console.timeLog("stringify json")

  const outFile = `${args[0]}.json`
  writeFileSync(outFile, json);
  console.log(`json written to ${outFile}`)
})();
```

The new engine is 3x faster and uses a third of the memory. Let me know if this works for you.

**Comment by nickbabcock** [comment]
Fixed in latest release (v0.3.x)
